tests/targets/test_twitter.py

```diff
@@ -117,19 +121,48 @@ def setUp(self):
 
     def test_publish(self):
         self.subject.publish()
-        self.api.PostUpdate.assert_called_once_with(self.subject.text())
+        text, reimbursement_image = self.subject.tweet_data()
+        self.api.PostUpdate.assert_called_once_with(
+            media=reimbursement_image, status=text)
         dict_representation = dict(self.subject)
         self.database.posts.insert_one.assert_called_once_with(
             dict_representation)
 
-    def test_text(self):
+    def test_tweet_data(self):
         message = (
             '🚨Gasto suspeito de Dep. @DepEduardoCunha (RJ). '
             'Você pode me ajudar a verificar? '
             'https://jarbas.serenata.ai/layers/#/documentId/10 '
             '#SerenataDeAmor na @CamaraDeputados'
         )
-        self.assertEqual(message, self.subject.text())
+        self.assertEqual(
+            (message, None), self.subject.tweet_data())
         self.reimbursement['twitter_profile'] = None
         with self.assertRaises(ValueError):
-            self.subject.text()
+            self.subject.tweet_data()
+
+    def test_tweet_text(self):
+        message = (
+            '🚨Gasto suspeito de Dep. @DepEduardoCunha (RJ). '
+            'Você pode me ajudar a verificar? '
+            'https://jarbas.serenata.ai/layers/#/documentId/10 '
+            '#SerenataDeAmor na @CamaraDeputados'
+        )
+        self.assertEqual(message, self.subject.tweet_text())
+
+    def test_camara_image_url(self):
+        url = 'http://www.camara.gov.br/cota-parlamentar/documentos/publ/10/2015/10.pdf'
+        self.assertEqual(url, self.subject.camara_image_url())
+
+    @mock.patch('whistleblower.targets.twitter.urllib.request.urlopen')
+    def test_tweet_image_success(self, urlopen_mock):
+        with open('tests/fixtures/10.pdf', 'rb') as mock_response:
+            urlopen_mock.return_value = BytesIO(mock_response.read())
+        self.assertIsInstance(self.subject.tweet_image(), BufferedReader)
```

Comment:
Regarding this test (lines 157-161): I haven't ran this code, but I might help with two comments – let me know if they are helpful ; )

**The mock**

Simplifying the source code, it reads:

```python
response = urllib.request.urlopen(self.camara_image_url())
image_bin = Image(file=response).make_blob('png')
```

So basically what you want to is whatever `response` means in the `Image` call. You want it to be a file object I guess. Thus you might not need to write the contents to `BytesIO` and directly use the result of the `open` context manager:

```python
with open('tests/fixtures/10.pdf', 'rb') as mock_response:
    urlopen_mock.return_value = mock_response
    self.assertIsInstance(self.subject.tweet_image(), BufferedReader)
```

Note that the assertion happens within the `with` block.

**The error**

The traceback tells me that there's a higher chance that the error is **not** related to the `urlopen` mock _per se_, but to the object it returns. Let's dig in the last few lines of the error:

```
File "/home/travis/build/okfn-brasil/whistleblower/whistleblower/targets/twitter.py", line 194, in tweet_image
    image_bin = Image(file=response).make_blob('png')
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/wand/image.py", line 2740, in __init__
    self.read(file=file, resolution=resolution)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/wand/image.py", line 2822, in read
    self.raise_exception()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/wand/resource.py", line 222, in raise_exception
    raise e
wand.exceptions.DelegateError: Postscript delegate failed `/tmp/magick-QbjrFeu8': No such file or directory @ error/pdf.c/ReadPDFImage/677
```

This tells us that when you call, in the test `Image(file=response).make_blob('png')`, internally `wand` tries to read this file with `self.read(file=file, resolution=resolution)`. This reading attempt fails raising an exception that tells us there's no file `/tmp/magick-QbjrFeu8`.

Not sure what this file is, but as you're mocking the `urlopen` response with `BytesIO` (which is something in memory, not in the file system) this might be the real problem. So I'd try to do as I suggested earlier because the fixture is in the file system. Alternatively you might need to create a file for that.